services/QuillLMS/spec/support/shared_contexts/snapshots/snapshots_cte.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+RSpec.shared_context 'Snapshots CTE' do
+  # BigQuery doesn't support CTE aliases with period in the name.
+  # So, we need to remove the 'lms.' and 'special.' prefixes from 'FROM' and 'JOIN' clauses
+  let(:query) { pg_query.gsub(/(FROM|JOIN)\s+(?:lms|special)\.(\w+)/, '\1 \2') }
+
+  let(:bq_query) do
+    <<-SQL
+      WITH
+        #{cte}
+      #{query}
+    SQL
+  end
+
+  let(:results) { QuillBigQuery::Runner.execute(bq_query) }
+
+  def cte
+    cte_table_collections
+      .reject { |collection| collection.empty? }
+      .map { |collection| "#{cte_alias(collection)} AS ( #{cte_query(collection)} )" }
+      .join(",\n\t")
```

Comment:
This is a pure style thing, but I think that this becomes safer, and maybe slightly more useful by extension, if we don't make strong assumptions that each collection is complete, or that there aren't multiple collections of the same model type.  Imagine that `cte_table_collections` is `[[classroom1, classroom2], [school], [edge_case_classroom]]`.  I think that you can handle that with Ruby's `group_by`:
```ruby
cte_table_collections
  .flatten
  .group_by { |record| record.class.table_name }
  .map do |table_name, collection|
    "#{table_name} AS ( #{collection.map { |record| [:SELECT, cte_select_clause(record)].join(' ') } } )"
  end.join(" UNION ALL \n")
```
Only, you know, cleaner than that.  I think that being able to be relatively imprecise with what goes into the `cte_table_collections` (or whatever we end up calling it) value will empower later developers to get simpler value out of this toolchain.

Reply:
Good idea with group_by.  I think this framework much easier to add records in.